### ISSUE TITLE
support userfaultfd?

### ISSUE THREAD

**Issue by fergald** [rr-debugger/rr#2852]
Chrome when running on chromeos uses the `userfaultfd` syscall and `rr` fails with

```
[FATAL /usr/local/google/home/fergal/github/rr-debugger/rr/src/record_syscall.cc:5846:rec_process_syscall_arch()]
 (task 1715052 (rec:1715052) at time 16963)
 -> Assertion `t->regs().syscall_result_signed() == -syscall_state.expect_errno' failed to hold. Expected ENOSYS for 'userfaultfd' but got result -22 (errno EINVAL); execution of syscall unsupported by rr
```

Relevant [Chrome src](https://source.chromium.org/chromium/chromium/src/+/main:chromeos/memory/userspace_swap/userspace_swap.cc;l=288;drc=53e082033509eb3b234a08af6d1d6048e43c7e03) is here.

Can we get `rr` to support this syscall? Thanks.

**Comment by rocallahan** [comment]
I didn't even know rr worked on ChromeOS!

**Comment by fergald** [comment]
Not sure if it works on "real" chromeos however it's possible to build and run a chromeos build of chrome on vanilla linux and run it (and test suites in particular). I don't have a chromeos device handy to tell you whether `rr` will work.

**Comment by fergald** [comment]
And also, just disabling the call to `userfaultfd` allows the test suite to run successfully, so I think supporting this syscall is all that's blocking the run-chromeos-tests-on-linux use case.

**Comment by rocallahan** [comment]
Real support for `userfaultfd` is pretty hard :-(.

AFAIK there aren't any explicit notifications we can observe to detect a thread waiting on a userfault or being woken up after a userfault has been serviced. But we need to detect these sleeps and wakeups so that our scheduler doesn't try to schedule a thread that won't make progress because it's waiting on a userfault.

Here are some tools we do have: using a `FileMonitor` we can observe and potentially override all tracee reads, writes and ioctls on a userfaultfd. We can also get a duplicate of the userfaultfd file descriptor into the rr process and some do things with it, e.g. check whether there are pending events. We can't easily read from it though, because there is no way to "put data back" so the application can see it. We might be able to replace it with a socket in the tracee and proxy operations on that socket to the real userfaultfd but that might be complicated and fragile.

A key question is how would we interrupt a running task that encounters a userfault. One approach would be to use `F_SETOWN_EX` and `F_SETSIG` to arrange for a signal to be delivered to the tracee when there is a pending fault event for the userfaultfd (similar to what we do with desched signals in blocking syscalls). Unfortunately a major problem is that if another thread faults on the same page before the first fault has been serviced, no new fault event is generated, so this approach will fail :-(. We could try to work around that by mprotecting PROT_NONE any pages that we detect have pending userfaults. To get the page address, we would have to read the userfault event, which means we'd have to do the proxying I mentioned above. This also requires machinery to temporarily mprotect pages during recording, invisibly handle the signals generated by faults on those pages, and do all this without interfering with normal operation of the tracee.

Alternatively, we might consider arming the desched event notification during the entire execution of a thread in an address space where user faulting is enabled. The problem then is, when a task is descheduled, how would we determine there was a userfault, and on which page (so that wakeups wake the correct task)? I don't have an answer to those questions.

There is another set of issues around faults generated in system calls. Userfaultfd for faults in system calls is "deprecated" and disabled for non-privileged code in newer distros/kernels, but it appears this ChromeOS code is not using `UFFD_USER_MODE_ONLY` and privileged code is still an issue. With userfaultfd I think *any* system call that reads or writes userspace memory can block on a userfault. This means a lot of system calls that we currently treat as not-possibly-blocking have to be treated as potentially blocking, at least when userfaultfds are present. This wouldn't be so bad, except some of those system calls get very special treatment (e.g. `clone()`, which can write `parent_tid`), where it would be extremely difficult to support context switching during that special treatment :-(. Maybe we can work around it by prefaulting-in relevant pages, but this adds complexity and fragility.

Another related issue is userfaults that happen during syscallbuf processing. Currently syscallbuf relies on context switches not happening except possibly during a blocking buffered syscall. If syscallbuf code execution triggers a userfault, we have no choice but to context switch to make forward progress :-(.

**Comment by rocallahan** [comment]
We could disable syscall buffering while userfaultfds are active, but that means a big slowdown for programs that use this.

**Comment by rocallahan** [comment]
> Here are some tools we do have: using a FileMonitor we can observe and potentially override all tracee reads, writes and ioctls on a userfaultfd.

To read from a userfaultfd without replacing it with a socket in the tracee, we could potentially override `read` etc to return buffered data that we already read and examined. However in general we would also have to override `select`, `poll`, `epoll` etc to act as if the fd is readable. This is also complicated :-(.

**Comment by rocallahan** [comment]
If supporting userfaultfd using existing kernel APIs is as hard as I think it is, here are some alternatives:
a) have userfaultfd return ENOSYS. It seems to me that might work for this ChromeOS code.
b) ask for (and, probably, implement ourselves) a first-class notification API in the kernel. Then it might be tractable to support `UFFD_USER_MODE_ONLY` at least.
c) do nothing

I think probably for now I will do a) and start a conversation around b) on LKML.

**Comment by rocallahan** [comment]
I posted https://lkml.org/lkml/2021/5/10/2607, proposing a new perf event `PERF_COUNT_SW_USERFAULTS` with the associated page address retrievable using `PERF_SAMPLE_ADDR` in the perf event ring buffer.

If that existed, then we could arm that event whenever userfaults are enabled for a task, and use `F_SETOWN_EX` to raise a signal in the tracee task when it stops on a userfault. We would then get the page address from the perf event and track which tasks are waiting for which pages. We would observe application ioctls that replace pages and wake tasks, and determine which tasks are no longer waiting. I think this would handle the `UFFD_USER_MODE_ONLY` case without much trouble. (We would just die if syscallbuf triggered a userfault, but that would be an application bug I guess.)

**Comment by fergald** [comment]
78611c3 worked for me. Thanks! I'll see if I can get someone involved in chromeos and userfault to chip in on this thread.

**Comment by dbaron** [comment]
Also just leaving a cross-reference to the [chromium-dev thread](https://groups.google.com/a/chromium.org/g/chromium-dev/c/c6Rp0vI4Fok/m/JNExqg0HAwAJ).  (Note that the timestamps that show up in that view are when the messages were accepted by the list moderator, which can be substantially later than when they're sent.)

**Comment by rocallahan** [comment]
Thanks. And I'm excited to see that chaos mode worked!
